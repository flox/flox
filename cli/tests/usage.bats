#! /usr/bin/env bats
# -*- mode: bats; -*-
# ============================================================================ #
#
# Test flox (no subcommand) command
#
# ---------------------------------------------------------------------------- #

load test_support.bash

# ---------------------------------------------------------------------------- #

# ---------------------------------------------------------------------------- #

setup() {
  common_test_setup
}
teardown() {
  common_test_teardown
}

@test "f1: simplify flox 'no command' info" {
  run "$FLOX_BIN"
  assert_success
  # FLOX_VERSION is set by the `flox run` command
  # and thus deviates from the expected version.
  assert_output --regexp - << EOF
flox version \d+.\d+.\d+-.+

Usage: flox OPTIONS \(init|activate|search|install|\.\.\.\) \[--help\]

Use "flox --help" for full list of commands and more information

First time\? Create an environment with "flox init"
EOF
}

@test "f?: 'flox --help' has 0 exit code" {
  run "$FLOX_BIN"
  assert_success
}

@test "f2: command grouping changes 1: 'Local Development Commands' listed in order" {
  run --separate-stderr "$FLOX_BIN" --help
  line=4
  assert_line -n "$line" --regexp '^Local Development Commands'
  line=$((line + 1))
  assert_line -n "$line" --regexp '^    init[ ]+[\w .,]+'
  line=$((line + 1))
  assert_line -n "$line" --regexp '^    activate[ ]+[\w .,]+'
  line=$((line + 1))
  assert_line -n "$line" --regexp '^    search[ ]+[\w .,]+'
  line=$((line + 1))
  assert_line -n "$line" --regexp '^    show[ ]+[\w .,]+'
  line=$((line + 1))
  assert_line -n "$line" --regexp '^    install, i[ ]+[\w .,]+'
  line=$((line + 1))
  assert_line -n "$line" --regexp '^    uninstall[ ]+[\w .,]+'
  line=$((line + 1))
  assert_line -n "$line" --regexp '^    edit[ ]+[\w .,]+'
  line=$((line + 1))
  assert_line -n "$line" --regexp '^    list[ ]+[\w .,]+'
  line=$((line + 1))
  assert_line -n "$line" --regexp '^    delete[ ]+[\w .,]+'
}

@test "f3: command grouping changes 2: 'Sharing Commands' listed in order" {
  run "$FLOX_BIN" --help
  line=14
  assert_line -n "$line" --regexp '^Sharing Commands'
  line=$((line + 1))
  assert_line -n "$line" --regexp '^    push[ ]+[\w .,]+'
  line=$((line + 1))
  assert_line -n "$line" --regexp '^    pull[ ]+[\w .,]+'
}

@test "f5: command grouping changes 3: move lesser used or not polished commands to 'Additional Commands' section with help tip." {
  run "$FLOX_BIN" --help
  assert_output --partial - << EOF
Additional Commands. Use "flox COMMAND --help" for more info
    update, upgrade, config, auth
EOF
}

@test "f6: remove stability from flox --help command: Only show stability for commands that support it" {
  run "$FLOX_BIN" --help
  refute_output --partial "--stability"
}
