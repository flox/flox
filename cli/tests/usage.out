usage:
    flox [(-h|--help)] [--version] [--prefix]

general commands:
    flox [(-v|--verbose)] [--debug] <command> [<args>]
    ----
    flox channels [--json]
         - list channel subscriptions
    flox subscribe [<name> [<url>]]
         - subscribe to channel URL
    flox unsubscribe [<name>]
         - unsubscribe from channel
    flox search [(-c,--channel) <channel>]... [(-l|--long)|--json] [--refresh] [<regex>[@<semver-range>]]
         - search packages in subscribed channels
    flox config [--list] [--reset [--confirm]] \
                [--set <arg> <value>] [--setNumber <arg> <value>] \
                [--delete <arg>]
         - configure user parameters
    flox gh - access to the gh CLI
    flox (envs|environments) - list all available environments

environment commands:
    flox <command> [(-e|--environment) <env>] [<args>]
    ----
    flox activate - activate environment:
        in current shell: eval "$(flox activate)"
        in subshell: flox activate
        for command: flox activate -- <command> <args>
    flox list [--out-path] [--json]
         - list installed packages
    flox create - create environment
    flox install - install a package into an environment
    flox (rm|remove) - remove packages from an environment
    flox upgrade [--force]
         - upgrade packages using their most recent flake
    flox edit - edit declarative form of an environment
    flox import - import a tar created with 'flox export' as a new generation
    flox export - export environment for use with 'flox import'
    flox history [--oneline] [--json]
         - show all versions of an environment
    flox generations [--json]
         - list environment generations with contents
    flox rollback - roll back to the previous generation of an environment
    flox switch-generation - switch to a specific generation of an environment
    flox delete [--force] [--origin]
         - remove all data pertaining to an environment
    flox push [--force] [-m|--main]
         - send environment metadata to remote registry
    flox pull [--force] [-m|--main] [--no-render]
         - pull and render environment from remote registry
    flox git - access to the git CLI for floxmeta repository

development commands:
    flox [--stability (stable|staging|unstable)] \
         [(-d|--date) <date_string>] <command> [<args>]
    ----
    flox init - initialize flox expressions for current project
    flox build - build package from current project
    flox eval - evaluate a Nix expression
    flox develop - launch development shell for current project
    flox print-dev-env - print shell code that can be sourced by bash to reproduce the development environment
    flox run - run app from current project
    flox shell - run a shell in which the current project is available
    flox publish [--build-repo <URL>] [--channel-repo <URL>] \
                 [--upload-to <URL>] [--download-from <URL>] \
                 [--render-path <dir>] [--key-file <file>] \
                 [(-A|--attr) <package>] [--publish-system <system>]
         - build and publish project to flox channel
